File: 17.letter-combinations-of-a-phone-number.cs
using System;
using System.Collections.Generic;

public class Solution17 {

    public IList<string> LetterCombinations(string digits) {
        IList<string> results = new List<string>();
        if (string.IsNullOrEmpty(digits))
        {
            return results;
        }
        String[] mapping = new String[] { "0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
        Queue<string> q = new Queue<string>();
        q.Enqueue("");
        for (int i = 0; i < digits.Length; ++i)
        {
            string str = mapping[digits[i] - '0'];
            while (q.Peek().Length == i)
            {
                var s = q.Dequeue();

                foreach (var c in str)
                {
                    q.Enqueue(s + c);
                }
            }
        }

        
        while(q.Count >0)
        {
            results.Add(q.Dequeue());
        }
        return results;
    }

    public IList<string> LetterCombinations2(string digits) {
        IList<string> results = new List<string>();
        if(string.IsNullOrEmpty(digits)) return results;

        // Can also use Backtracking to solve this by track the current stringbuilder.

        return results;
    }
}

File: 17_letter-combinations-of-a-phone-number.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        phone = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl', 
             '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}
        if not digits:
            return []

        queue = []
        queue.append('')

        for i in range(len(digits)):
            while len(queue[0]) == i:
                tmp = queue.pop(0)
                for j in range(len(phone[digits[i]])):
                    queue.append(tmp + phone[digits[i]][j])

        return queue